### PR TITLE
add support for m/henvcfgh and mconfigptr CSRs 

### PR DESCRIPTION
**Pull request by liweiwei90**

### CODE REVIEW

File: riscv/csrs.cc [hunk 1]
```diff
@@ -467,7 +467,10 @@ mstatus_csr_t::mstatus_csr_t(processor_t* const proc, const reg_t addr):
       | (proc->extension_enabled_const('S') && (proc->get_const_xlen() != 32) ? set_field((reg_t)0, MSTATUS_SXL, xlen_to_uxl(proc->get_const_xlen())) : 0)
 
 #ifdef RISCV_ENABLE_DUAL_ENDIAN
-      | (proc->get_mmu()->is_target_big_endian() ? MSTATUS_UBE | MSTATUS_SBE | MSTATUS_MBE : 0)
+      | (proc->get_mmu()->is_target_big_endian() ? (proc->extension_enabled_const('U') ? MSTATUS_UBE : 0)
```
Comment (scottj97) [inline riscv/csrs.cc]: b570b4493768ca6f563b08e7a0c2f7c90fcb6739 looks like 4 separate commits to me

File: riscv/csrs.cc [hunk 2]
```diff
@@ -858,7 +862,7 @@ satp_csr_t::satp_csr_t(processor_t* const proc, const reg_t addr):
 void satp_csr_t::verify_permissions(insn_t insn, bool write) const {
   base_atp_csr_t::verify_permissions(insn, write);
   if (get_field(state->mstatus->read(), MSTATUS_TVM))
-    require(state->prv >= PRV_M);
+    require(state->prv == PRV_M);
```
Comment (scottj97) [inline riscv/csrs.cc:888]: This should be its own commit with a message such as: "Clarify permission check since prv can never be larger than PRV_M" (if I'm understanding this correctly?)

File: riscv/processor.cc [hunk 3]
```diff
@@ -388,8 +388,11 @@ void state_t::reset(processor_t* const proc, reg_t max_isa)
                              (proc->extension_enabled(EXT_ZICBOZ) ? MENVCFG_CBZE : 0) |
                              (proc->extension_enabled(EXT_SVPBMT) ? MENVCFG_PBMTE : 0);
   const reg_t menvcfg_init = (proc->extension_enabled(EXT_SVPBMT) ? MENVCFG_PBMTE : 0);
-  csrmap[CSR_MENVCFG] = menvcfg = std::make_shared<masked_csr_t>(proc, CSR_MENVCFG, menvcfg_mask, menvcfg_init);
-  if (xlen == 32) csrmap[CSR_MENVCFGH] = std::make_shared<rv32_high_csr_t>(proc, CSR_MENVCFGH, -1, menvcfg);
+  menvcfg = std::make_shared<masked_csr_t>(proc, CSR_MENVCFG, menvcfg_mask, menvcfg_init);
```
Comment (scottj97) [inline riscv/processor.cc]: Would it be okay to never initialize `menvcfg` if there's no U-mode? I can't see that any code would access it in that case.

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/csrs.cc (hunk 1 above)
Comment (liweiwei90) [reply riscv/csrs.cc]: Yeah. It contains four point of u/s related check.

File: riscv/csrs.cc (hunk 2 above)
Comment (liweiwei90) [reply riscv/csrs.cc:888]: OK. I'll do this later.

File: riscv/processor.cc (hunk 3 above)
Comment (liweiwei90) [reply riscv/processor.cc]: Leaving menvcfg here is for compatibility. menvcfg will be used to check whether PBMT is supported (in walk and s2xlate) , such as in s2xlate:
`      bool pbmte = proc->get_state()->menvcfg->read() & MENVCFG_PBMTE;`
If we never initialize menvcfg if there's no U-mode, this function of PBMT will be affected.

Comment (scottj97) [reply riscv/processor.cc]: If there's no U-mode then there's no S-mode either (right?) so paging will never be enabled. The code in `s2xlate()` and friends should never be executed.

Comment (liweiwei90) [reply riscv/processor.cc]: Yeah. I agree.  I'll update this.

File: riscv/csrs.cc [hunk 4]
```diff
@@ -862,7 +862,7 @@ satp_csr_t::satp_csr_t(processor_t* const proc, const reg_t addr):
...
```
Comment (scottj97) [inline riscv/csrs.cc:888]: The commit message of "fix permission check" is misleading, even alarming, because it makes it sound like there was a bug with this CSR's permission checks. But this commit doesn't actually change any behavior. It clarifies a confusing line of code, that's all.

File: riscv/csrs.cc [hunk 5]
```diff
@@ -467,7 +467,10 @@ mstatus_csr_t::mstatus_csr_t(processor_t* const proc, const reg_t addr):
       | (proc->extension_enabled_const('S') && (proc->get_const_xlen() != 32) ? set_field((reg_t)0, MSTATUS_SXL, xlen_to_uxl(proc->get_const_xlen())) : 0)
 
 #ifdef RISCV_ENABLE_DUAL_ENDIAN
-      | (proc->get_mmu()->is_target_big_endian() ? MSTATUS_UBE | MSTATUS_SBE | MSTATUS_MBE : 0)
+      | (proc->get_mmu()->is_target_big_endian() ? (proc->extension_enabled_const('U') ? MSTATUS_UBE : 0)
+                                                 | (proc->extension_enabled_const('S') ? MSTATUS_SBE : 0)
+                                                 | MSTATUS_MBE
+                                                 : 0)
```
Comment (scottj97) [inline riscv/csrs.cc]: The syntax is awkward here because of the ternary operator. What if you made a separate const `big_endian_bits` first, and then use that to make this ternary into a single line again?

File: riscv/csrs.cc (hunk 4 above)
Comment (liweiwei90) [reply riscv/csrs.cc:888]: OK. I'll update this.

File: riscv/csrs.cc (hunk 5 above)
Comment (liweiwei90) [reply riscv/csrs.cc]: OK.

File: riscv/csrs.cc [hunk 6]
```diff
@@ -465,24 +465,28 @@ bool sstatus_proxy_csr_t::unlogged_write(const reg_t val) noexcept {
...
+    const reg_t big_endian_bits = (proc->extension_enabled_const('U') ? MSTATUS_UBE : 0)
+                                | (proc->extension_enabled_const('S') ? MSTATUS_SBE : 0)
+                                | MSTATUS_MBE;
+#endif
+    val = 0
+          | (proc->extension_enabled_const('U') && (proc->get_const_xlen() != 32) ? set_field((reg_t)0, MSTATUS_UXL, xlen_to_uxl(proc->get_const_xlen())) : 0)
+          | (proc->extension_enabled_const('S') && (proc->get_const_xlen() != 32) ? set_field((reg_t)0, MSTATUS_SXL, xlen_to_uxl(proc->get_const_xlen())) : 0)
 #ifdef RISCV_ENABLE_DUAL_ENDIAN
-      | (proc->get_mmu()->is_target_big_endian() ? MSTATUS_UBE | MSTATUS_SBE | MSTATUS_MBE : 0)
+          | (proc->get_mmu()->is_target_big_endian() ? big_endian_bits : 0)
```
Comment (scottj97) [inline riscv/csrs.cc]: Sorry, I did not realize when I requested this change that it was part of an initializer list. `val` really ought to be initialized in the initializer list, not in the constructor body.

You could use a [delegating constructor](https://stackoverflow.com/a/17375163/202201) though perhaps that's too much. Maybe there's a way to fit this all in one expression that would be easier to read than the original?

Comment (liweiwei90) [reply riscv/csrs.cc]: Can we move all this into a separate function, and initialize the val with this function in the initializer list? just like the way for sstatus_write_mask of base_status_csr_t:
`sstatus_write_mask(compute_sstatus_write_mask())`

Comment (scottj97) [reply riscv/csrs.cc]: Yes, good idea!

File: riscv/csrs.cc [hunk 7]
```diff
@@ -497,6 +490,20 @@ bool mstatus_csr_t::unlogged_write(const reg_t val) noexcept {
   return true;
 }
 
+reg_t mstatus_csr_t::compute_mstatus_initial_value() const noexcept {
+#ifdef RISCV_ENABLE_DUAL_ENDIAN
```
Comment (scottj97) [inline riscv/csrs.cc]: I don't the either of these `#ifdef`s are needed here. `is_target_big_endian()` will always be false if this macro is undefined, and it doesn't hurt to always calculate these values.

**Review by scottj97 — CHANGES_REQUESTED** [review]
Commit 022e5ac2afffce2a3bf2eb7e288a000c9aaf343d combines a refactoring (extract new function) with functional changes. Please split into two so the functional changes are more apparent.

File: riscv/csrs.cc (hunk 7 above)
Comment (liweiwei90) [reply riscv/csrs.cc]: Yeah, I think so and remove the similar  #ifdef RISCV_ENABLE_DUAL_ENDIAN in other functions.

File: fesvr/htif.h [hunk 8]
```diff
@@ -31,26 +31,18 @@ class htif_t : public chunked_memif_t
 
   template<typename T> inline T from_target(target_endian<T> n) const
   {
-#ifdef RISCV_ENABLE_DUAL_ENDIAN
```
Comment (scottj97) [inline fesvr/htif.h:34]: This change I'm not sure about. It is functionally correct, but might cause a performance degradation. `from_target()` and `to_target()` are called frequently.

This is probably why there was a compile-time define in the first place, versus only the dynamic runtime endianness choice.

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: fesvr/htif.h (hunk 8 above)
Comment (liweiwei90) [reply fesvr/htif.h:34]: It seems only be used by to_host/from_host related read/write.

Comment (scottj97) [reply fesvr/htif.h:34]: ??

`from_target()` is used in the `load_func()` macro which means every load calls it. `to_target()` is used in the `store_func()` macro which means every store calls it.

Comment (liweiwei90) [reply fesvr/htif.h:34]: I think load/store_func use the  from/to_target in mmu.h itself not this one.

Comment (scottj97) [reply fesvr/htif.h:34]: Yes, my mistake.

**Review by scottj97 — APPROVED** [review]

**Comment by aswaterman** [comment]
Thanks @liweiwei90 and @scottj97!
